pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,777 @@
+func (c *ExternalNodeController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting controller", "name", controllerName)
+	defer klog.InfoS("Shutting down controller", "name", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.externalNodeListerSynced) {
+		klog.Error("Failed to wait for syncing ExternalNodes cache")
+		return
+	}
+
+	if err := c.reconcile(); err != nil {
+		klog.ErrorS(err, "Failed to reconcile ExternalNodes")
+		return
+	}
+
+	go wait.Until(c.worker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeAdd(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	if en.Name != c.nodeName || en.Namespace != c.namespace {
+		return
+	}
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode ADD event", "ExternalNode", klog.KObj(en))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeUpdate(oldObj interface{}, newObj interface{}) {
+	oldEn := oldObj.(*v1alpha1.ExternalNode)
+	newEn := newObj.(*v1alpha1.ExternalNode)
+	if newEn.Name != c.nodeName || newEn.Namespace != c.namespace {
+		return
+	}
+	if reflect.DeepEqual(oldEn.Spec.Interfaces, newEn.Spec.Interfaces) {
+		klog.InfoS("Skip enqueuing ExternalNode UPDATE event as no changes for interfaces", "ExternalNode", klog.KObj(newEn))
+		return
+	}
+	key, _ := keyFunc(newEn)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode UPDATE event", "ExternalNode", klog.KObj(newEn))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeDelete(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	if en.Name != c.nodeName || en.Namespace != c.namespace {
+		return
+	}
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode DELETE event", "ExternalNode", klog.KObj(en))
+}
+
+func (c *ExternalNodeController) reconcile() error {
+	klog.InfoS("Reconciliation for controller", "name", controllerName)
+	if err := c.reconcileHostUplinkFlows(); err != nil {
+		return fmt.Errorf("failed to reconcile host uplink flows %v", err)
+	}
+	if err := c.reconcileExternalNodeInterfaces(); err != nil {
+		return fmt.Errorf("failed to reconcile ExternalNode interfaces %v", err)
+	}
+	if err := c.reconcilePolicyBypassFlows(); err != nil {
+		return fmt.Errorf("failed to reconcile reserved flows %v", err)
+	}
+	klog.Info("ExternalNodeController finishes reconciliation")
+	return nil
+}
+
+func (c *ExternalNodeController) reconcileHostUplinkFlows() error {
+	hostIfaces := c.ifaceStore.GetInterfacesByType(interfacestore.ExternalEntityInterface)
+	for _, hostIface := range hostIfaces {
+		if err := c.ofClient.InstallVMUplinkFlows(hostIface.InterfaceName, hostIface.OVSPortConfig.OFPort, hostIface.UplinkPort.OFPort); err != nil {
+			klog.ErrorS(err, "Failed to install openflow entries to forward packet between uplink and host interface", "hostInterface", hostIface.InterfaceName)
```

Comment:
packet -> packets

But I feel you can simply say "Failed to install host uplink flows".